File: app/seed.py
"""Seed initial data for MyQuang restaurant."""
from datetime import datetime
from passlib.context import CryptContext
from .database import SessionLocal, engine
from .models import Base, Floor, Room, Table, MenuCategory, MenuItem
from .models import PaymentMethod, ExpenseCategory, Role, User, Setting

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def seed():
    Base.metadata.create_all(bind=engine)
    db = SessionLocal()

    try:
        # Skip if already seeded
        if db.query(Floor).count() > 0:
            print("Database already seeded. Skipping.")
            return

        print("Seeding database...")

        # ── Roles ──
        role_admin = Role(name="admin", description="Chủ quán - Toàn quyền",
            permissions='["all"]')
        role_cashier = Role(name="cashier", description="Thu ngân",
            permissions='["order.view","order.create","order.edit","payment.create","payment.view","menu.view"]')
        role_waiter = Role(name="waiter", description="Phục vụ",
            permissions='["order.view","order.create","order.edit","menu.view"]')
        db.add_all([role_admin, role_cashier, role_waiter])
        db.flush()

        # ── Users ──
        admin = User(role_id=role_admin.id, username="admin",
            full_name="Chủ quán MyQuang",
            password_hash=pwd_context.hash("admin123"),
            pin_code="1234")
        cashier = User(role_id=role_cashier.id, username="thungan",
            full_name="Thu ngân",
            password_hash=pwd_context.hash("thungan123"),
            pin_code="3456")
        waiter = User(role_id=role_waiter.id, username="phucvu",
            full_name="Phục vụ",
            password_hash=pwd_context.hash("phucvu123"),
            pin_code="4567")
        db.add_all([admin, cashier, waiter])
        db.flush()

        # ── Floors ──
        f1 = Floor(name="Tầng 1", sort_order=1)
        f2 = Floor(name="Tầng 2", sort_order=2)
        db.add_all([f1, f2])
        db.flush()

        # ── Rooms (Tầng 2) ──
        room_a = Room(floor_id=f2.id, name="Phòng A", capacity=16, sort_order=1)
        room_b = Room(floor_id=f2.id, name="Phòng B", capacity=16, sort_order=2)
        db.add_all([room_a, room_b])
        db.flush()

        # ── Tables ──
        tables = []
        # Tầng 1: 6 bàn
        for i in range(1, 7):
            tables.append(Table(
                floor_id=f1.id, room_id=None,
                code=f"T1-B{i:02d}", name=f"Bàn {i:02d}",
                capacity=4 if i <= 4 else 6
            ))
        # Phòng A: 4 bàn
        for i in range(1, 5):
            tables.append(Table(
                floor_id=f2.id, room_id=room_a.id,
                code=f"T2-PA-B{i:02d}", name=f"Bàn A{i:02d}",
                capacity=4
            ))
        # Phòng B: 4 bàn
        for i in range(1, 5):
            tables.append(Table(
                floor_id=f2.id, room_id=room_b.id,
                code=f"T2-PB-B{i:02d}", name=f"Bàn B{i:02d}",
                capacity=4
            ))
        db.add_all(tables)

        # ── Menu Categories ──
        cat_main = MenuCategory(name="Món chính", sort_order=1)
        cat_starter = MenuCategory(name="Khai vị", sort_order=2)
        cat_drink = MenuCategory(name="Đồ uống", sort_order=3)
        cat_dessert = MenuCategory(name="Tráng miệng", sort_order=4)
        db.add_all([cat_main, cat_starter, cat_drink, cat_dessert])
        db.flush()

        # ── Menu Items ──
        items = [
            MenuItem(category_id=cat_main.id, name="Cơm sườn nướng", base_price=65000, sort_order=1),
            MenuItem(category_id=cat_main.id, name="Cơm gà xào sả ớt", base_price=60000, sort_order=2),
            MenuItem(category_id=cat_main.id, name="Bún bò Huế", base_price=55000, sort_order=3),
            MenuItem(category_id=cat_main.id, name="Phở bò tái nạm", base_price=60000, sort_order=4),
            MenuItem(category_id=cat_main.id, name="Mì xào hải sản", base_price=70000, sort_order=5),
            MenuItem(category_id=cat_starter.id, name="Gỏi cuốn (2 cuốn)", base_price=35000, sort_order=1),
            MenuItem(category_id=cat_starter.id, name="Chả giò (4 cái)", base_price=40000, sort_order=2),
            MenuItem(category_id=cat_starter.id, name="Soup cua", base_price=45000, sort_order=3),
            MenuItem(category_id=cat_drink.id, name="Cà phê đen đá", base_price=25000, sort_order=1),
            MenuItem(category_id=cat_drink.id, name="Cà phê sữa đá", base_price=30000, sort_order=2),
            MenuItem(category_id=cat_drink.id, name="Trà đào cam sả", base_price=45000, sort_order=3),
            MenuItem(category_id=cat_drink.id, name="Nước ngọt lon", base_price=20000, sort_order=4),
            MenuItem(category_id=cat_drink.id, name="Bia Tiger lon", base_price=35000, sort_order=5),
            MenuItem(category_id=cat_drink.id, name="Nước suối", base_price=15000, sort_order=6),
            MenuItem(category_id=cat_dessert.id, name="Chè đậu xanh", base_price=25000, sort_order=1),
            MenuItem(category_id=cat_dessert.id, name="Kem ba màu", base_price=30000, sort_order=2),
        ]
        db.add_all(items)

        # ── Payment Methods ──
        db.add_all([
            PaymentMethod(name="Tiền mặt", description="Khách trả tiền mặt"),
            PaymentMethod(name="Chuyển khoản", description="Chuyển khoản / quét QR"),
            PaymentMethod(name="Momo", description="Ví điện tử Momo"),
        ])

        # ── Expense Categories ──
        db.add_all([
            ExpenseCategory(name="Nguyên vật liệu"),
            ExpenseCategory(name="Điện - Nước"),
            ExpenseCategory(name="Lương nhân viên"),
            ExpenseCategory(name="Thuê mặt bằng"),
            ExpenseCategory(name="Sửa chữa - Bảo trì"),
            ExpenseCategory(name="Marketing"),
            ExpenseCategory(name="Chi phí khác"),
        ])

        # ── Settings ──
        db.add_all([
            Setting(key="restaurant_name", value="Quán Ăn MyQuang", description="Tên quán"),
            Setting(key="restaurant_address", value="123 Đường ABC, TP.HCM", description="Địa chỉ"),
            Setting(key="restaurant_phone", value="0901234567", description="Số điện thoại"),
            Setting(key="tax_rate", value="0", description="Thuế suất (%)"),
            Setting(key="service_charge_rate", value="0", description="Phí phục vụ (%)"),
            Setting(key="currency", value="VND", description="Đơn vị tiền tệ"),
            Setting(key="receipt_footer_note", value="Cảm ơn quý khách đã dùng bữa tại MyQuang! ❤️", description="Ghi chú phiếu thu"),
            Setting(key="bank_account", value="", description="Số tài khoản ngân hàng"),
            Setting(key="bank_name", value="", description="Tên ngân hàng"),
            Setting(key="bank_owner", value="", description="Tên chủ tài khoản"),
        ])

        db.commit()
        print("✅ Database seeded successfully!")
        print("   Tài khoản: admin / admin123")
        print("   Thu ngân: thungan / thungan123")
        print("   Phục vụ: phucvu / phucvu123")

    except Exception as e:
        db.rollback()
        print(f"❌ Seed failed: {e}")
        raise
    finally:
        db.close()


if __name__ == "__main__":
    seed()

File: src/seed.js
const bcrypt = require('bcryptjs');
const { db, q } = require('./db');

function seed() {
  const existing = q.get('SELECT COUNT(*) as cnt FROM floors');
  if (existing && existing.cnt > 0) {
    console.log('Database already seeded. Skipping.');
    return;
  }

  console.log('Seeding database...');

  db.exec('BEGIN');
  try {
    // ── Roles ──
    const roleAdmin = q.run(
      `INSERT INTO roles (name, description, permissions) VALUES (?,?,?)`,
      'admin', 'Chủ quán - Toàn quyền', '["all"]'
    );
    const roleCashier = q.run(
      `INSERT INTO roles (name, description, permissions) VALUES (?,?,?)`,
      'cashier', 'Thu ngân', '["order.view","order.create","order.edit","payment.create","payment.view","menu.view"]'
    );
    const roleWaiter = q.run(
      `INSERT INTO roles (name, description, permissions) VALUES (?,?,?)`,
      'waiter', 'Phục vụ', '["order.view","order.create","order.edit","menu.view"]'
    );

    // ── Users ──
    q.run(
      `INSERT INTO users (role_id, username, full_name, password_hash, pin_code) VALUES (?,?,?,?,?)`,
      roleAdmin.lastInsertRowid, 'admin', 'Chủ quán MyQuang',
      bcrypt.hashSync('admin123', 10), '1234'
    );
    q.run(
      `INSERT INTO users (role_id, username, full_name, password_hash, pin_code) VALUES (?,?,?,?,?)`,
      roleCashier.lastInsertRowid, 'thungan', 'Thu ngân',
      bcrypt.hashSync('thungan123', 10), '3456'
    );
    q.run(
      `INSERT INTO users (role_id, username, full_name, password_hash, pin_code) VALUES (?,?,?,?,?)`,
      roleWaiter.lastInsertRowid, 'phucvu', 'Phục vụ',
      bcrypt.hashSync('phucvu123', 10), '4567'
    );

    // ── Floors ──
    const f1 = q.run(`INSERT INTO floors (name, sort_order) VALUES (?,?)`, 'Tầng 1', 1);
    const f2 = q.run(`INSERT INTO floors (name, sort_order) VALUES (?,?)`, 'Tầng 2', 2);

    // ── Rooms ──
    const roomA = q.run(
      `INSERT INTO rooms (floor_id, name, capacity, sort_order) VALUES (?,?,?,?)`,
      f2.lastInsertRowid, 'Phòng A', 16, 1
    );
    const roomB = q.run(
      `INSERT INTO rooms (floor_id, name, capacity, sort_order) VALUES (?,?,?,?)`,
      f2.lastInsertRowid, 'Phòng B', 16, 2
    );

    // ── Tables: Tầng 1 ──
    for (let i = 1; i <= 5; i++) {
      q.run(
        `INSERT INTO tables (floor_id, room_id, code, name, capacity) VALUES (?,?,?,?,?)`,
        f1.lastInsertRowid, null,
        `T1-B${String(i).padStart(2,'0')}`,
        `Bàn ${String(i).padStart(2,'0')}`,
        i <= 4 ? 4 : 6
      );
    }
    // ── Tables: Phòng A ──
    for (let i = 1; i <= 4; i++) {
      q.run(
        `INSERT INTO tables (floor_id, room_id, code, name, capacity) VALUES (?,?,?,?,?)`,
        f2.lastInsertRowid, roomA.lastInsertRowid,
        `T2-PA-B${String(i).padStart(2,'0')}`,
        `Bàn A${String(i).padStart(2,'0')}`, 4
      );
    }
    // ── Tables: Phòng B ──
    for (let i = 1; i <= 4; i++) {
      q.run(
        `INSERT INTO tables (floor_id, room_id, code, name, capacity) VALUES (?,?,?,?,?)`,
        f2.lastInsertRowid, roomB.lastInsertRowid,
        `T2-PB-B${String(i).padStart(2,'0')}`,
        `Bàn B${String(i).padStart(2,'0')}`, 4
      );
    }

    // ── Sentinel: Mang về (không hiện trên sơ đồ bàn) ──
    q.run(
      `INSERT INTO tables (floor_id, code, name, capacity, status, is_active, is_takeaway)
       VALUES (?, 'MV', 'Mang về', 0, 'available', 1, 1)`,
      f1.lastInsertRowid
    );

    // ── Menu Categories ──
    const catMiQuang  = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Mì Quảng Truyền Thống', 1);
    const catBunMam   = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Bún Mắm Nêm', 2);
    const catCuon     = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Món Cuốn', 3);
    const catAnKem    = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Món Ăn Kèm', 4);
    const catComGa    = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Cơm Gà Hội An', 5);
    const catDrink    = q.run(`INSERT INTO menu_categories (name, sort_order) VALUES (?,?)`, 'Giải Khát', 6);

    // ── Menu Items ──
    const menuItems = [
      // Mì Quảng Truyền Thống
      [catMiQuang.lastInsertRowid, 'Mì Quảng Đặc Biệt',   80000, 1],
      [catMiQuang.lastInsertRowid, 'Mì Quảng Bò',         65000, 2],
      [catMiQuang.lastInsertRowid, 'Mì Quảng Ếch',        60000, 3],
      [catMiQuang.lastInsertRowid, 'Mì Quảng Gà',         55000, 4],
      [catMiQuang.lastInsertRowid, 'Mì Quảng Tôm Thịt',   55000, 5],
      [catMiQuang.lastInsertRowid, 'Mì Quảng Cá Lóc',     55000, 6],
      // Bún Mắm Nêm
      [catBunMam.lastInsertRowid,  'Bún Mắm Nêm Thập Cẩm Đặc Biệt', 75000, 1],
      [catBunMam.lastInsertRowid,  'Bún Mắm Heo Quay (Thường)',     55000, 2],
      [catBunMam.lastInsertRowid,  'Bún Mắm Heo Quay (Đặc Biệt)',   65000, 3],
      [catBunMam.lastInsertRowid,  'Bún Mắm Thịt Luộc (Thường)',    45000, 4],
      [catBunMam.lastInsertRowid,  'Bún Mắm Thịt Luộc (Đặc Biệt)',  55000, 5],
      // Món Cuốn
      [catCuon.lastInsertRowid,    'Bánh Tráng Cuốn Heo Quay',  80000, 1],
      [catCuon.lastInsertRowid,    'Bánh Tráng Cuốn Thịt Luộc', 70000, 2],
      // Món Ăn Kèm
      [catAnKem.lastInsertRowid,   'Bánh Bột Lọc (Đĩa 6 cái)',  30000, 1],
      [catAnKem.lastInsertRowid,   'Chả Ram Tôm Bình Định',     35000, 2],
      // Cơm Gà Hội An
      [catComGa.lastInsertRowid,   'Cơm Gà Hội An (Phần Nhỏ)',      55000, 1],
      [catComGa.lastInsertRowid,   'Cơm Gà Hội An (Phần Đầy Đặn)',  65000, 2],
      // Giải Khát
      [catDrink.lastInsertRowid,   'Coca Cola',         15000, 1],
      [catDrink.lastInsertRowid,   'Pepsi',             15000, 2],
      [catDrink.lastInsertRowid,   'Trà Chanh / Quất',  15000, 3],
      [catDrink.lastInsertRowid,   'Trà Sâm Dứa',        5000, 4],
      [catDrink.lastInsertRowid,   'Nước Suối',         10000, 5],
    ];
    const catNames = {
      [catMiQuang.lastInsertRowid]: 'Mì Quảng Truyền Thống',
      [catBunMam.lastInsertRowid]: 'Bún Mắm Nêm',
      [catCuon.lastInsertRowid]: 'Món Cuốn',
      [catAnKem.lastInsertRowid]: 'Món Ăn Kèm',
      [catComGa.lastInsertRowid]: 'Cơm Gà Hội An',
      [catDrink.lastInsertRowid]: 'Giải Khát',
    };
    menuItems.forEach(([cat, name, price, sort]) => {
      const img = resolveMenuImage(name, catNames[cat] || '');
      q.run(
        `INSERT INTO menu_items (category_id, name, base_price, sort_order, image_url) VALUES (?,?,?,?,?)`,
        cat, name, price, sort, img
      );
    });

    const { ensureInventoryItems, splitCombinedDrink, syncRecipes } = require('./migrate-inventory');
    const { resolveMenuImage } = require('./lib/menu-stock-images');
    ensureInventoryItems();
    splitCombinedDrink();
    syncRecipes();

    // ── Payment Methods ──
    ['Tiền mặt', 'Chuyển khoản', 'Momo'].forEach(name => {
      q.run(`INSERT INTO payment_methods (name) VALUES (?)`, name);
    });

    // ── Expense Categories ──
    ['Nguyên vật liệu','Điện - Nước','Lương nhân viên','Thuê mặt bằng','Sửa chữa - Bảo trì','Marketing','Chi phí khác'].forEach(name => {
      q.run(`INSERT INTO expense_categories (name) VALUES (?)`, name);
    });

    // ── Settings ──
    const defaults = [
      ['restaurant_name',    'Quán Ăn MyQuang',          'Tên quán'],
      ['restaurant_address', '123 Đường ABC, TP.HCM',    'Địa chỉ'],
      ['restaurant_phone',   '0901234567',               'Số điện thoại'],
      ['tax_rate',           '0',                        'Thuế suất (%)'],
      ['service_charge_rate','0',                        'Phí phục vụ (%)'],
      ['currency',           'VND',                      'Đơn vị tiền tệ'],
      ['receipt_footer_note','Cảm ơn quý khách đã dùng bữa tại MyQuang!', 'Ghi chú phiếu thu'],
      ['bank_account',       '7167676767',               'Số tài khoản'],
      ['bank_name',          'MB',                       'Tên ngân hàng (mã VietQR)'],
      ['bank_owner',         'HO KINH DOANH E E.. MI QUANG QUE', 'Tên chủ tài khoản'],
    ];
    defaults.forEach(([key, value, desc]) => {
      q.run(`INSERT INTO settings (key, value, description) VALUES (?,?,?)`, key, value, desc);
    });

    db.exec('COMMIT');
    console.log('✅ Database seeded!');
    console.log('   admin / admin123');
    console.log('   thungan / thungan123');
    console.log('   phucvu / phucvu123');
  } catch (e) {
    db.exec('ROLLBACK');
    console.error('❌ Seed failed:', e.message);
    throw e;
  }
}

module.exports = { seed };
